### PR TITLE
Roman Cycle 9 Updates

### LINKED ISSUE #403: Add the new K short filter to Roman WFI
Add spectral information for the additional bandpass filter added to WFI in fall 2020.

### LINKED ISSUE #443: Update Roman Jitter value
New Jitter value should be changed from 0.014 arcsec per axis to 0.012 arcsec per axis

### LINKED ISSUE #465: WebbPSF-Roman Cycle 9 Update
WebbPSF Data:

- [x] Update new Zernike phase NOLL coefficients
    - [x] WFI - (Sheet Tab:  `WIM`)
    - [x] GRISM - (Sheet Tab:  `WIM-GRISM`)
    - [x] PRISM - (Sheet Tab:  `WIM-PRISM`)

- [x] New filter called the K filter (Sub-issue #403)

- [x] Filter bandpass info.

- [x] Update Pupil Masks
    - [x] Each SCA has its own copy 
    - [x] 4 types of masks (Prism and Grism have their own Pupils now)

WebbPSF Code:

- [x] Pupil controller 
- [x] Tests 
- [x] Docs
    - [x] Add the following to docs "The jitter for the Roman observatory is 0.012 arcsec per axes"

Checks:

- [x] Instrument layout diagram (currently outdated in the ref file)
- [x] Design numbers (have not been checked for a while)

### PR DESCRIPTION
**Pull request by ojustino**
Would resolve #465.

# Updates

#### `WFIPupilController`
- Added `pupil_file_formatters` attribute that will hopefully allow for easier updates when pupil filename structures change in future cycles.
- Created `update_pupil()` from the remains of the former `_update_pupil()`, `update_pupil_path()`, and `validate_pupil()`. This method points `WFI` to the appropriate pupil file and pupil mask string after the user changes `WFI`'s detector or filter. It is now the only standard approach for updating the pupil or pupil mask.
  - Added `_get_filter_mask()` to help match a given filter with its corresponding pupil mask.
- As such, locked `pupil_mask` and `pupil` so that direct assignment is no longer possible, though `_pupil` and `_pupil_mask` allow for indirect assignment in `update_pupil()`.
- Added `lock_pupil()` and `lock_pupil_mask()` methods to disable default pupil/mask selection behavior by toggling the new `_auto_pupil` and `_auto_pupil_mask` attributes to False, respectively. Instead, they lock `WFI` on the provided pupil file or pupil mask string. `unlock_pupil()` and `unlock_pupil_mask()` re-enable default behavior.
- Removed all (un)masked pupil-related attributes as there's no such distinction in the new pupil files.
- ~~`GRISM_FILTER`~~ `GRISM_FILTERS` and ~~`PRISM_FILTER`~~ `PRISM_FILTERS` are now tuples containing the new string labels for those filters in `filters.tsv` (including separate entries for grism's 0th and 1st degrees).

#### `WFI`
- Pointed  ~~`_aberrations_files`~~ `_aberration_files` attribute to Cycle 9's Zernike files.
  - Renamed ~~`_is_custom_aberrations`~~ `_is_custom_aberration` and ~~`_current_aberrations_file`~~ `_current_aberration_file`.
- Removed `_unmasked_pupil_path` and `_masked_pupil_path` since their corresponding attributes were deleted from `WFIPupilController`.
- Renamed ~~`override_aberrations()`~~ `lock_aberrations()` and ~~`reset_override_aberrations()`~~ `unlock_aberrations()`.
- Created new `lock_pupil()` and `lock_pupil_mask()` methods for advanced users who prefer to disable automatic pupil/mask selection and instead stick with a specific pupil file or mask, respectively. The corresponding `unlock_pupil()` and `unlock_pupil_mask()` methods return the class to its normal behavior. (These methods follow in the footsteps of `lock_aberrations()` and `unlock_aberrations()`.)
- As such, locked `pupil` and `pupil_mask` so that direct assignment is no longer possible. Both attributes continue to dynamically update on any change to detector or filter that necessitates it.
- Added `_update_pupil()` as a shortcut to `WFIPupilController.update_pupil()` used in the lock/unlock methods and the detector/filter setters.

#### `RomanInstrument`
- Updated `jitter_sigma` to 0.012 arcsec/axis to reflect [new simulations](https://roman.ipac.caltech.edu/sims/Param_db.html#telescope) (fixes #443).

#### Other
- Made corresponding changes to Roman test and documentation files.

## webbpsf-data-source
(`webbpsf-data/` ➡️ `webbpsf-data-source/` after #416)

### Out with the old...
- **Archived** Cycle 8 Zernike files
  - `webbpsf-data-source/WFI/wim_zernikes_cycle8.csv` ➡️ `webbpsf-data-source/WFI/sources/cycle8/wim_zernikes_cycle8.csv`
  - `webbpsf-data-source/WFI/wim_zernikes_cycle8_grism.csv` ➡️ `webbpsf-data-source/WFI/sources/cycle8/wim_zernikes_cycle8_grism.csv`
  - `webbpsf-data-source/WFI/wim_zernikes_cycle8_prism.csv` ➡️ `webbpsf-data-source/WFI/sources/cycle8/wim_zernikes_cycle8_prism.csv`
- **Archived** previous cycle's throughput-by-filter files (`webbpsf-data-source/WFI/filters/*_throughput.fits` ➡️ `webbpsf-data-source/WFI/sources/cycle8/filters/*_throughput.fits`)
- **Archived** previous cycle's pupil files (`webbpsf-data-source/WFI/pupils` ➡️ `webbpsf-data-source/WFI/sources/cycle8/new_pupils`)
- _(Previous cycle's post-grism/prism update version of `filters.tsv` already archived in `webbpsf-data-source/WFI/sources/cycle8_prism_grism/filters.tsv`)_
- _(Previous cycle's HST OPD file already archived in `webbpsf-data-source/WFI/sources/cycle8/new_upscaled_HST_OPD.fits)`_

### ...in with the new.

(To examine file processing procedures, see `webbpsf-data-source/WFI/sources/cycle9/Process_zern_and_filters.ipynb` for Zernikes and filter files or `webbpsf-data-source/WFI/sources/cycle9/Oversample_rescale_pupil_images_cycle_9.ipynb` for pupil files.)

- **Updated** Zernike files _(note change from `wim` to `wsm` in spectral mode filenames to better distinguish them from imaging mode)_
  - `webbpsf-data-source/WFI/sources/cycle9/wim_zernikes_cycle9.csv`
    - (Needs to be sorted by SCA for some reason?)
  - `webbpsf-data-source/WFI/sources/cycle9/wsm_zernikes_cycle9_grism.csv`
  - `webbpsf-data-source/WFI/sources/cycle9/wsm_zernikes_cycle9_prism.csv`
- **Added** new imaging filter and **renamed** prism/grism filters as described in next section (`webbpsf-data-source/WFI/sources/cycle9/filters.tsv`)
- **Created** filter throughput files (`webbpsf-data-source/WFI/sources/cycle9/filters/*_throughput.fits`)
- **Updated** OPD file to match new pupil images' pixel scale (`webbpsf-data-source/WFI/sources/cycle9/new_upscaled_HST_OPD.fits`)

The files and directories were "pre-archived" in `webbpsf-data-source/WFI/sources/cycle9/` and then copied to their appropriate locations in `webbpsf-data-source` (the renamed `upscaled_HST_OPD.fits`) or `webbpsf-data-source/WFI/` (the rest).

# Checks
- [] **Filter bandpasses**
  - [X] F062
  - [X] F087
  - [X] F106
  - [X] F129
  - [X] F146
  - [X] F158
  - [X] F184
  - [X] F213 (new, resolves #403)
  - [X] ~~P120~~ --> PRISM (newly named)
  - [X] ~~G150~~ --> GRISM0 (newly named and split)
  - [X] ~~G150~~ --> GRISM1 (newly named and split)

- [X] **Instrument layout** (no changes)

- [X] **Zernike coefficients**

- [X] **Pupil image compatibility**

# Notable Roman design changes

image

- As seen above, the wavelength range now covers 0.48 to 2.3 micrometers.
  - The new F213 filter adds coverage from 1.95 to 2.3 micrometers.
- The grism mode is now split into an undispersed zeroth order and dispersed first order.
- Sampling points are no longer four pixels from each corner.

### CODE REVIEW

File: webbpsf/roman.py
Comment (Skyhawk172) [inline webbpsf/roman.py:807]: It appears like this is doing the right thing, mainly because of the call to self._update_pupil(). The reset is based on the current values of wfi.filter and wfi.detector.

File: webbpsf/roman.py
Comment (Skyhawk172) [inline webbpsf/roman.py]: Same as above (line 815). The reset is based on the current values of `wfi.filter` and `wfi.detector`, and they seem to behave as expected.

File: webbpsf/roman.py
Comment (Skyhawk172) [inline webbpsf/roman.py]: I'll also defer to your best programmer judgment here.

File: webbpsf/roman.py
Comment (Skyhawk172) [inline webbpsf/roman.py]: I'll defer to your best programmer judgment here.

File: webbpsf/roman.py
Comment (Skyhawk172) [inline webbpsf/roman.py:508]: Limited understanding of the class here, so bear with me. I'm not sure if there's a use case where this particular method gets called. 

From what I see, the locking/unlocking of the pupil and pupil_mask is done on the WFI object, not the WFIPupilController object.  The former seems to be handling the unlocking as expected because it knows the values of the filter and detector, and resets them using `wfi._update_pupil()`. 

But WFIPupilController, by itself, has no knowledge of these values, so it cannot properly reset the pupil and pupil_mask.  When we unlock the pupil, all it does is set the `auto_lock` to `True`, leaving the pupil and pupil_mask values as they were.  

When instantiated, the pupil and pupil_mask are set to None - maybe that's what they should be reset to.

File: webbpsf/tests/test_roman.py
Comment (Skyhawk172) [inline webbpsf/tests/test_roman.py]: As I mentioned on Slack, I'm able to run `calc_psf()` with GRISM0, so I'm not sure what's happening here. This particular test is passing for me even when I remove your IF GRISM0 statement:

test_roman.py::test_WFI_filters PASSED                                                               [100%]                                                                                                                                                              


======================== PASSES ======================== 
------------------------------- Captured stdout call -----------------------
F062
F087
F106
F129
F146
F158
F184
PRISM
F213
GRISM0
GRISM1
====================== 1 passed in 33.75s ===================

**Review by Skyhawk172 — APPROVED** [review]

File: webbpsf/roman.py
Comment (ojustino) [reply webbpsf/roman.py:807]: Agreed, thanks for the sanity check.

File: webbpsf/roman.py
Comment (ojustino) [reply webbpsf/roman.py]: I don't think these are wrong, but a better alternative isn't coming to me. I'll leave it as is for the sake of time.

File: webbpsf/roman.py
Comment (ojustino) [reply webbpsf/roman.py:508]: You're right in saying that `WFIPupilController` has no knowledge of the current pupil or pupil mask value and that the actual assignment of the `pupil_mask` attribute happens in `WFI`. This note was a reminder to re-check the WFI version of `unlock_pupil_mask()` since it calls this one.

File: webbpsf/tests/test_roman.py
Comment (ojustino) [reply webbpsf/tests/test_roman.py]: This is the only comment that I'd say requires a follow up. `calc_psf()` with GRISM0 errors out for me even with a fresh install and a new environment, and I think the failure makes sense given what I commented in the code about `get_weights()` in poppy.

What do you get when you try the following (replacing the ellipsis with your path to the new data files)?
```
from astropy.io import fits
any(fits.getdata('.../WFI/filters/GRISM0_throughput.fits')['THROUGHPUT'] > .1)
```

If it's `False`, I'm not sure why poppy would accept the weights as valid.

File: docs/roman.rst
Comment (mperrin) [inline docs/roman.rst:76]: This text does not seem correct; the reference pixels don't vary. It's a fixed width of 4 pixels around all edges of all detectors. Rephrase to something like the following. 

```
The usable, photosensitive regions of the Wide Field Instrument's detectors are slightly smaller than their 4096 by 4096 pixel dimensions, because the outermost 4 rows and columns are reference pixels that are not sensitive to light. 
```

But that said, I don't understand entirely why this is relevant here for the PSF calculations.  This is a detector property, not something in the optical models.

File: webbpsf/roman.py
Comment (mperrin) [inline webbpsf/roman.py:803]: This is out of scope for this PR, but at some point it might be worth revisiting some of the complexity that's been added into the WFI classes, such as this lock and unlock pupil functionality. What is the use case for that? Is it still needed? Does anyone actually use this part of the code? I think it may be the case that some of the code added over time is no longer particularly useful, and could potentially be simplified out.

Comment (ojustino) [reply webbpsf/roman.py:803]: Robel and I spoke with Adric about this and Adric said it's useful in his workflows for the Roman ETC (and perhaps Pandeia). Robel also told me that there are some other, more intrepid users who like having this functionality.

**Comment by ojustino** [comment]
@mperrin figured out that the GRISM0 discrepancy happened because `poppy`'s `Instrument._get_weights()` works differently depending on whether or not `synphot` is present. `synphot` is currently optional but will become a requirement in WebbPSF 1.0. 

I've still left GRISM0 out of `test_WFI_filters()` to keep this pull request's tests passing. I'll create an issue to reverse that once the new release is out.

**Comment by mperrin** [comment]
@ojustino This looks good, apart from one quibble with some doc string text. We can fix that separately as part of other docs edits for release 1.0. I'm going to go ahead and merge this now. 

We ought to be able to add GRISM0 into the tests immediately, since once this is merged into develop the tests will run with synphot installed. So let's try immediately making a PR to reverse that right away after I merge this.

File: docs/roman.rst
Comment (ojustino) [reply docs/roman.rst:76]: I talked about it with Charles but will catch up with you about this.

**Pull request merged**
